src/dataframe/dataframe.jl

```diff
@@ -303,11 +276,16 @@ function Base.getindex(df::DataFrame, row_inds::AbstractVector, col_inds::Abstra
 end
 
 # df[:, SingleColumnIndex] => AbstractVector
+function Base.getindex(df::DataFrame, row_inds::Colon, col_ind::ColumnIndex)
+    selected_column = index(df)[col_ind]
+    copy(_columns(df)[selected_column][row_inds])
+end
+
 # df[:, MultiColumnIndex] => DataFrame
-function Base.getindex(df::DataFrame, row_ind::Colon, col_inds)
-    Base.depwarn("Indexing with colon as row will create a copy in the future. " *
-                 "Use `df[col_inds]` to get the columns without copying", :getindex)
-    df[col_inds]
+function Base.getindex(df::DataFrame, row_ind::Colon, col_inds::AbstractVector)
+    selected_columns = index(df)[col_inds]
+    new_columns = AbstractVector[copy(dv) for dv in _columns(df)[selected_columns]]
+    return DataFrame(new_columns, Index(_names(df)[selected_columns]))
```

Comment:
Isn't there a more efficient way to create a subset of an `Index` than recreating one from scratch? Just wondering.

Reply:
An alternative I can think of would be to `deepcopy` the old index and remove columns from it, but it does not seem that it will be faster neither when `col_inds` is small nor when it is large relative to the original number of columns. So I would leave it as is for now.